### ISSUE TITLE
rc-buffer doesn't recognize Rc<String> in some cases

### ISSUE THREAD

**Issue by m1el** [rust-lang/rust-clippy#8059]
### Summary

rc-buffer doesn't recognize `Rc<Buffer>` when used in function body.

If a local function variable has `Rc<String>` type, the lint should trigger, but it does not.

### Lint Name

rc-buffer

### Reproducer

I tried this code:

```rust
use std::collections::BTreeMap;
use std::rc::Rc;

fn main() {
    let mut bt = BTreeMap::<Rc<String>, u64>::new();
    let key: Rc<String> = Rc::from("foo".to_string());
    bt.insert(key, 12);
    println!("{:?}", bt.get(&"foo".to_string()));
}
```

I expected to see this happen:
clippy displays rc-buffer lint for lines 4 and 5 and suggests to use `Rc<str>` instead of `Rc<String>`.

Instead, this happened:
clippy returns no lint messages.

### Version

```text
rustc 1.59.0-nightly (48a5999fc 2021-12-01)
binary: rustc
commit-hash: 48a5999fceeea84a8971634355287faa349909d4
commit-date: 2021-12-01
host: x86_64-pc-windows-msvc
release: 1.59.0-nightly
LLVM version: 13.0.0
```
